lib/active_merchant/billing/gateways/fat_zebra.rb

```diff
@@ -125,16 +127,42 @@ def add_creditcard(post, creditcard, options = {})
       def add_extra_options(post, options)
         extra = {}
         extra[:ecm] = '32' if options[:recurring]
-        extra[:cavv] = options[:cavv] || options.dig(:three_d_secure, :cavv) if options[:cavv] || options.dig(:three_d_secure, :cavv)
-        extra[:xid] = options[:xid] || options.dig(:three_d_secure, :xid) if options[:xid] || options.dig(:three_d_secure, :xid)
-        extra[:sli] = options[:sli] || options.dig(:three_d_secure, :eci) if options[:sli] || options.dig(:three_d_secure, :eci)
         extra[:name] = options[:merchant] if options[:merchant]
         extra[:location] = options[:merchant_location] if options[:merchant_location]
         extra[:card_on_file] = options.dig(:extra, :card_on_file) if options.dig(:extra, :card_on_file)
         extra[:auth_reason]  = options.dig(:extra, :auth_reason) if options.dig(:extra, :auth_reason)
+
+        unless options[:three_d_secure].present?
```

Comment:
This validation and those extra values are handled in the #add_three_ds method, the difference, is the place you are getting the field values.

❓ Question:

Why do you have two places to populate the sli, xid, and the cvv extra fields? (add_extra_options and add_three_ds methods), Is it possible to include the if options[:sli] conditional in the conditionals for add_3ds?

Reply:
@gasb150 I found that someone basically added some fields related to 3DS in the options hash, so to prevent damage on someonelse work I leaved there just in case.